torch/csrc/autograd/variable.h

```diff
@@ -152,8 +153,28 @@ struct TORCH_API Variable : public at::Tensor {
 
   // NOTE: Assignment operators to Tensor come for free from the constructors.
 
-  const at::Tensor& data() const noexcept;
-  at::Tensor& data() noexcept;
+  /// NOTE: This is similar to the legacy `.data()` function on `Variable`, and is intended
+  /// to be used from functions that need to access the `Variable`'s equivalent `Tensor`
+  /// (i.e. `Tensor` that shares the same storage and tensor metadata with the `Variable`).
+  ///
+  /// One notable difference with the legacy `.data()` function is that changes to the
+  /// returned `Tensor`'s tensor metadata (e.g. sizes / strides / storage / storage_offset)
+  /// will not update the original `Variable`, due to the fact that this function
+  /// shallow-copies the `Variable`'s underlying TensorImpl.
+  at::Tensor tensor_data() const noexcept;
```

Comment:
@gchanan We couldn't keep the non-const version of the legacy `data()` (i.e. `at::Tensor& data()`), because the underlying implementation of `tensor_data()` has to shallow-copy the TensorImpl, and the shallow copy itself is a temporary object, which we can't have a non-const reference to (https://stackoverflow.com/questions/13826897/why-not-non-const-reference-to-temporary-objects).